File: Sources/CodexMonitor/Monitor.swift
import AppKit
import Combine
import Foundation

@MainActor
final class QuotaMonitor: ObservableObject {
    static let shared = QuotaMonitor()

    struct Entry: Identifiable {
        var profile: Profile
        var isActive: Bool
        var snapshot: UsageSnapshot?
        /// True when `snapshot` came from the on-disk cache or a previous successful fetch.
        var isStale: Bool
        var error: String?
        var isFetching: Bool
        /// Newest account-wide limit seen in a session rollout (real-time, same data as the app).
        var liveEvent: LiveRateLimitEvent?
        /// Newest per-model limit events, keyed by `limit_id` (e.g. Spark).
        var liveExtras: [String: LiveRateLimitEvent] = [:]

        var id: String { profile.id }

        /// True when the live event should be displayed instead of the API snapshot: it is newer
        /// than the last fetch, or it is recent and the API still reports *less* usage for the
        /// same window (the usage endpoint can lag the per-response headers by a little).
        var showsLive: Bool {
            guard let live = liveEvent else { return false }
            guard let snap = snapshot else { return true }
            if live.timestamp > snap.fetchedAt { return true }
            guard snap.fetchedAt.timeIntervalSince(live.timestamp) < 120,
                  let lp = live.primary, let ap = snap.usage.rateLimit?.primaryWindow else { return false }
            let sameWindow = abs((lp.resetAt ?? 0) - (ap.resetAt ?? 0)) < 60
            return sameWindow && lp.usedPercent > ap.usedPercent
        }

        /// What the card shows: the live event when it is newer than the API snapshot.
        var displayRateLimit: RateLimit? {
            if showsLive, let live = liveEvent { return live.asRateLimit }
            return snapshot?.usage.rateLimit
        }

        /// Additional (per-model) limits, patched with live events when those are newer.
        var displayAdditionalLimits: [AdditionalRateLimit] {
            let base = snapshot?.usage.additionalRateLimits ?? []
            let fetched = snapshot?.fetchedAt ?? .distantPast
            return base.map { extra in
                guard let key = extra.meteredFeature, let live = liveExtras[key], live.timestamp > fetched else { return extra }
                var patched = extra
                patched.rateLimit = live.asRateLimit
                return patched
            }
        }

        /// Timestamp and origin of the numbers currently displayed.
        var dataSource: (live: Bool, at: Date)? {
            if showsLive, let live = liveEvent { return (true, live.timestamp) }
            if let s = snapshot { return (false, s.fetchedAt) }
            return nil
        }

        var primaryWindow: RateLimitWindow? {
            displayRateLimit?.windows.first
        }

        /// The tightest constraint right now (lowest remaining %) across the main windows.
        var tightestWindow: RateLimitWindow? {
            displayRateLimit?.windows.min { $0.remainingPercent < $1.remainingPercent }
        }

        var resetCreditCount: Int? {
            snapshot?.resetCredits?.availableCount ?? snapshot?.usage.rateLimitResetCredits?.availableCount
        }
    }

    @Published private(set) var entries: [Entry] = []
    @Published private(set) var lastRefresh: Date?
    @Published private(set) var isRefreshing = false
    @Published private(set) var recentMessages: [String] = []
    /// Bumped once a minute so relative times and countdowns re-render.
    @Published private(set) var clockTick = 0
    static let intervalChoices = [15, 30, 60, 120, 300]

    @Published var refreshIntervalSeconds: Int {
        didSet {
            UserDefaults.standard.set(refreshIntervalSeconds, forKey: "refreshIntervalSeconds")
            scheduleTimer()
        }
    }
    /// When the server rejects an access token (401), exchange the refresh token once, like Codex
    /// itself does. A rejected token is already useless everywhere, so this cannot break a copy
    /// on another machine; it only restores this one. Default on.
    @Published var autoRefreshOn401: Bool {
        didSet { UserDefaults.standard.set(autoRefreshOn401, forKey: "autoRefreshOn401") }
    }
    private var lastAutoRefreshAttempt: [String: Date] = [:]

    let store = ProfileStore.shared
    private var client: UsageClient
    private var refreshTimer: Timer?
    private var pollTimer: Timer?
    private var tickTimer: Timer?
    private var lastSignature = ""
    private var started = false
    /// Reset-credit details change rarely; re-fetch them at most this often per account.
    private let creditsMaxAge: TimeInterval = 180
    private var lastCreditsFetch: [String: Date] = [:]
    /// Set after an HTTP 429; automatic polling pauses until then (manual refresh still works).
    private var backoffUntil: Date?
    /// One tailer per `sessions` directory being watched, keyed by path.
    private var tailers: [String: RolloutTailer] = [:]
    private var liveTimer: Timer?
    /// Rollout events older than the last change of ~/.codex/auth.json may belong to a
    /// previously active account, so they are ignored for the main sessions directory.
    private var mainAuthModified: Date = .distantPast

    private init() {
        let d = UserDefaults.standard
        var seconds = d.integer(forKey: "refreshIntervalSeconds")
        if seconds <= 0 {
            // Migrate the pre-1.1 minutes setting; otherwise default to 30 s.
            let minutes = d.integer(forKey: "refreshIntervalMinutes")
            seconds = minutes > 0 ? min(minutes * 60, 300) : 30
        }
        refreshIntervalSeconds = seconds
        autoRefreshOn401 = d.object(forKey: "autoRefreshOn401") as? Bool ?? true
        client = UsageClient(baseURL: ProfileStore.shared.chatGPTBaseURL())
    }

    // MARK: Lifecycle

    func start() {
        guard !started else { return }
        started = true
        if DemoData.enabled {
            entries = DemoData.entries()
            lastRefresh = Date()
            tickTimer = Timer.scheduledTimer(withTimeInterval: 60, repeats: true) { [weak self] _ in
                Task { @MainActor in self?.clockTick += 1 }
            }
            return
        }
        store.ensureDirectories()
        reloadProfiles()
        Task { await refreshAll() }
        scheduleTimer()
        pollTimer = Timer.scheduledTimer(withTimeInterval: 10, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.pollFiles() }
        }
        tickTimer = Timer.scheduledTimer(withTimeInterval: 60, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.clockTick += 1 }
        }
        pollLive()
        liveTimer = Timer.scheduledTimer(withTimeInterval: 2, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.pollLive() }
        }
        NSWorkspace.shared.notificationCenter.addObserver(
            forName: NSWorkspace.didWakeNotification, object: nil, queue: .main
        ) { [weak self] _ in
            Task { @MainActor in
                try? await Task.sleep(nanoseconds: 4_000_000_000)
                await self?.refreshAll()
            }
        }
    }

    private func scheduleTimer() {
        refreshTimer?.invalidate()
        let interval = TimeInterval(max(10, refreshIntervalSeconds))
        refreshTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
            Task { @MainActor in await self?.refreshAll(automatic: true) }
        }
        refreshTimer?.tolerance = min(5, interval / 10)
    }

    /// Refresh now unless data is younger than `seconds` (used when the panel becomes visible).
    func refreshIfStale(seconds: TimeInterval) async {
        if let last = lastRefresh, Date().timeIntervalSince(last) < seconds { return }
        await refreshAll()
    }

    // MARK: Profiles

    /// Re-scan disk. Existing snapshots are carried over by account id so the UI does not flicker.
    func reloadProfiles() {
        if DemoData.enabled { return }
        let main = store.loadMain()
        var profiles = store.loadProfiles()

        if let main = main, let msg = store.adopt(main: main, into: profiles) {
            log(msg)
            profiles = store.loadProfiles()
        }

        var previous: [String: Entry] = [:]
        for e in entries {
            if let acct = e.profile.accountId, previous[acct] == nil { previous[acct] = e }
        }

        mainAuthModified = main?.modified ?? .distantPast

        func carryOver(_ e: inout Entry) {
            guard let acct = e.profile.accountId else { return }
            if let prev = previous[acct] {
                e.snapshot = prev.snapshot
                e.isStale = prev.isStale
                e.error = prev.error
                e.liveEvent = prev.liveEvent
                e.liveExtras = prev.liveExtras
            } else if let cached = loadCache(accountId: acct) {
                e.snapshot = cached
                e.isStale = true
            }
        }

        var newEntries: [Entry] = []
        let mainAccount = main?.accountId
        for p in profiles {
            var e = Entry(profile: p, isActive: mainAccount != nil && p.accountId == mainAccount,
                          snapshot: nil, isStale: false, error: nil, isFetching: false)
            carryOver(&e)
            newEntries.append(e)
        }
        // ~/.codex/auth.json holding an account that is not stored as a profile yet.
        if let main = main, main.auth != nil, !profiles.contains(where: { $0.accountId == main.accountId }) {
            var e = Entry(profile: main, isActive: true, snapshot: nil, isStale: false, error: nil, isFetching: false)
            carryOver(&e)
            newEntries.insert(e, at: 0)
        }
        // Active account first, then alphabetical.
        newEntries.sort { a, b in
            if a.isActive != b.isActive { return a.isActive }
            return a.profile.name.localizedStandardCompare(b.profile.name) == .orderedAscending
        }
        entries = newEntries
        lastSignature = currentSignature()
    }

    private func currentSignature() -> String {
        var parts: [String] = []
        if let m = store.loadMain() { parts.append(m.signature) }
        parts.append(contentsOf: store.loadProfiles().map { $0.signature })
        return parts.joined(separator: "\n")
    }

    private func pollFiles() {
        let sig = currentSignature()
        guard sig != lastSignature else { return }
        log(L.authChanged)
        reloadProfiles()
        Task { await refreshAll() }
    }

    // MARK: Live rollout events

    /// Directories whose rollouts describe this entry's account: the shared `~/.codex/sessions`
    /// for whichever account is active, plus the profile's own `sessions` (from `codex-acct run`).
    private func sessionsDirs(for entry: Entry) -> [(dir: URL, isMain: Bool)] {
        var dirs: [(URL, Bool)] = []
        if entry.isActive {
            dirs.append((store.codexHome.appendingPathComponent("sessions"), true))
        }
        if !entry.profile.isMain {
            let own = entry.profile.directory.appendingPathComponent("sessions")
            if FileManager.default.fileExists(atPath: own.path) { dirs.append((own, false)) }
        }
        return dirs
    }

    private func pollLive() {
        let now = Date()
        for idx in entries.indices {
            for (dir, isMain) in sessionsDirs(for: entries[idx]) {
                let tailer: RolloutTailer
                if let existing = tailers[dir.path] {
                    tailer = existing
                } else {
                    tailer = RolloutTailer(sessionsDir: dir)
                    tailers[dir.path] = tailer
                }
                let latest = tailer.pollLatest(now: now)
                if let ev = latest.main, !(isMain && ev.timestamp < mainAuthModified),
                   ev.timestamp > (entries[idx].liveEvent?.timestamp ?? .distantPast) {
                    entries[idx].liveEvent = ev
                    let used = ev.primary?.usedPercent ?? ev.secondary?.usedPercent ?? -1
                    NSLog("[CodexMonitor] live rate-limit event for %@: used %.1f%% at %@ (applied %.1fs later)",
                          entries[idx].profile.displayName, used, ISO8601.format(ev.timestamp), now.timeIntervalSince(ev.timestamp))
                }
                for (id, ev) in latest.extras where !(isMain && ev.timestamp < mainAuthModified) {
                    if ev.timestamp > (entries[idx].liveExtras[id]?.timestamp ?? .distantPast) {
                        entries[idx].liveExtras[id] = ev
                    }
                }
            }
        }
    }

    // MARK: Fetching

    /// `automatic` refreshes respect the 429 back-off; manual ones always go through.
    func refreshAll(automatic: Bool = false) async {
        guard !isRefreshing, !DemoData.enabled else { return }
        if automatic, let until = backoffUntil, until > Date() { return }
        isRefreshing = true
        defer { isRefreshing = false }
        client = UsageClient(baseURL: store.chatGPTBaseURL())
        let ids = entries.map { $0.id }
        await withTaskGroup(of: Void.self) { group in
            for id in ids {
                group.addTask { @MainActor in await self.refresh(entryId: id) }
            }
        }
        lastRefresh = Date()
    }

    func refresh(entryId: String, allowAutoRefresh: Bool = true) async {
        guard let idx = entries.firstIndex(where: { $0.id == entryId }) else { return }
        let profile = entries[idx].profile
        guard profile.isChatGPTAuth, let token = profile.accessToken else {
            entries[idx].error = profile.loadError ?? UsageError.noToken.localizedDescription
            return
        }
        if profile.identity?.accessTokenExpired == true {
            if allowAutoRefresh {
                switch await autoRefreshIfAllowed(entryId: entryId) {
                case .refreshed:
                    await refresh(entryId: entryId, allowAutoRefresh: false)
                    return
                case .failed:
                    return
                case .skipped:
                    break
                }
            }
            let exp = profile.identity?.accessTokenExpiry.map { Fmt.dateTime.string(from: $0) } ?? "?"
            if let i = entries.firstIndex(where: { $0.id == entryId }) {
                entries[i].error = L.tokenExpiredNeedRefresh(exp)
            }
            return
        }
        entries[idx].isFetching = true
        let client = self.client
        let acct = profile.accountId
        let previousCredits = entries[idx].snapshot?.resetCredits
        let creditsKey = acct ?? entryId
        let creditsDue = previousCredits == nil
            || Date().timeIntervalSince(lastCreditsFetch[creditsKey] ?? .distantPast) > creditsMaxAge
        do {
            let usage = try await client.fetchUsage(accessToken: token, accountId: acct)
            var credits = previousCredits
            if creditsDue {
                if let fresh = try? await client.fetchResetCredits(accessToken: token, accountId: acct) {
                    credits = fresh
                    lastCreditsFetch[creditsKey] = Date()
                }
            }
            let snap = UsageSnapshot(fetchedAt: Date(), usage: usage, resetCredits: credits)
            if let i = entries.firstIndex(where: { $0.id == entryId }) {
                entries[i].snapshot = snap
                entries[i].isStale = false
                entries[i].error = nil
                entries[i].isFetching = false
            }
            if let acct = acct { saveCache(snap, accountId: acct) }
        } catch {
            if case UsageError.http(let code, _) = error, code == 429 {
                backoffUntil = Date().addingTimeInterval(120)
                log(L.rateLimitedBackoff)
            }
            if let ue = error as? UsageError, ue.isAuthFailure, allowAutoRefresh {
                if let i = entries.firstIndex(where: { $0.id == entryId }) { entries[i].isFetching = false }
                switch await autoRefreshIfAllowed(entryId: entryId) {
                case .refreshed:
                    await refresh(entryId: entryId, allowAutoRefresh: false)
                    return
                case .failed:
                    return  // error text already set to "session revoked"
                case .skipped:
                    break
                }
            }
            if let i = entries.firstIndex(where: { $0.id == entryId }) {
                entries[i].error = (error as? UsageError)?.localizedDescription ?? error.localizedDescription
                entries[i].isFetching = false
                if entries[i].snapshot != nil { entries[i].isStale = true }
            }
        }
    }

    private enum AutoRefreshOutcome { case refreshed, failed, skipped }

    /// One refresh-token exchange per account per 10 minutes, only when enabled. On success the
    /// new tokens are written to the profile (and to ~/.codex/auth.json when that profile is the
    /// active one) and the profiles are reloaded so the retry uses them.
    private func autoRefreshIfAllowed(entryId: String) async -> AutoRefreshOutcome {
        guard autoRefreshOn401, let idx = entries.firstIndex(where: { $0.id == entryId }) else { return .skipped }
        let entry = entries[idx]
        let profile = entry.profile
        guard let rt = profile.auth?.tokens?.refreshToken, !rt.isEmpty else { return .skipped }
        let key = profile.accountId ?? profile.id
        if let last = lastAutoRefreshAttempt[key], Date().timeIntervalSince(last) < 600 { return .skipped }
        lastAutoRefreshAttempt[key] = Date()
        do {
            let refreshed = try await client.refreshTokens(refreshToken: rt)
            try store.applyRefreshedTokens(refreshed, to: profile.authURL)
            if entry.isActive, !profile.isMain {
                try store.applyRefreshedTokens(refreshed, to: store.mainAuthURL)
            }
            log(L.autoRefreshed(profile.displayName))
            reloadProfiles()
            return .refreshed
        } catch {
            log(L.autoRefreshFailed(profile.displayName, error.localizedDescription))
            if let i = entries.firstIndex(where: { $0.id == entryId }) {
                entries[i].error = L.sessionRevoked
                entries[i].isFetching = false
                if entries[i].snapshot != nil { entries[i].isStale = true }
            }
            return .failed
        }
    }

    // MARK: Actions

    func activate(entryId: String) {
        guard let entry = entries.first(where: { $0.id == entryId }), !entry.profile.isMain else { return }
        let main = store.loadMain()
        let profiles = store.loadProfiles()
        do {
            let msgs = try store.activate(entry.profile, currentMain: main, profiles: profiles)
            msgs.forEach { log($0) }
        } catch {
            log(L.switchFailed(error.localizedDescription))
        }
        reloadProfiles()
        Task { await refreshAll() }
    }

    func saveMainAsProfile(named name: String) {
        do {
            let p = try store.saveMainAsProfile(named: name)
            log(L.savedAsProfile(p.name))
        } catch {
            log(L.saveFailed(error.localizedDescription))
        }
        reloadProfiles()
        Task { await refreshAll() }
    }

    /// Explicit, user-confirmed token refresh. Writes the profile's auth.json and, when that
    /// profile is the active one, `~/.codex/auth.json` as well so both copies stay in lock-step.
    func refreshTokens(entryId: String) async {
        guard let idx = entries.firstIndex(where: { $0.id == entryId }) else { return }
        let profile = entries[idx].profile
        guard let rt = profile.auth?.tokens?.refreshToken, !rt.isEmpty else {
            log(L.noRefreshToken(profile.displayName))
            return
        }
        entries[idx].isFetching = true
        do {
            let refreshed = try await client.refreshTokens(refreshToken: rt)
            try store.applyRefreshedTokens(refreshed, to: profile.authURL)
            if entries[idx].isActive, !profile.isMain {
                try store.applyRefreshedTokens(refreshed, to: store.mainAuthURL)
            }
            log(L.tokensRefreshed(profile.displayName))
        } catch {
            log(L.refreshFailed(profile.displayName, error.localizedDescription))
        }
        if let i = entries.firstIndex(where: { $0.id == entryId }) { entries[i].isFetching = false }
        reloadProfiles()
        await refreshAll()
    }

    // MARK: Derived

    var activeEntry: Entry? { entries.first { $0.isActive } }

    var menuBarTitle: String {
        guard let e = activeEntry else { return entries.isEmpty ? "-" : "?" }
        if let w = e.tightestWindow {
            let pct = Fmt.percent(w.remainingPercent)
            return e.isStale ? "\(pct)*" : pct
        }
        if e.error != nil { return "!" }
        return "..."
    }

    // MARK: Cache

    private func cacheURL(accountId: String) -> URL {
        store.cacheDir.appendingPathComponent("usage-\(accountId).json")
    }

    private func loadCache(accountId: String) -> UsageSnapshot? {
        guard let data = try? Data(contentsOf: cacheURL(accountId: accountId)) else { return nil }
        let dec = JSONDecoder()
        dec.dateDecodingStrategy = .iso8601
        return try? dec.decode(UsageSnapshot.self, from: data)
    }

    private func saveCache(_ snap: UsageSnapshot, accountId: String) {
        let enc = JSONEncoder()
        enc.dateEncodingStrategy = .iso8601
        enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let data = try? enc.encode(snap) else { return }
        try? store.writeAtomically(data, to: cacheURL(accountId: accountId))
    }

    // MARK: Log

    /// UI-originated events (copy/export) that should appear in the panel log.
    func note(_ message: String) { log(message) }

    private func log(_ message: String) {
        let stamp = Fmt.dateTime.string(from: Date())
        recentMessages.insert("\(stamp) \(message)", at: 0)
        if recentMessages.count > 20 { recentMessages.removeLast(recentMessages.count - 20) }
        NSLog("[CodexMonitor] %@", message)
    }
}

File: codex_monitor/monitor.py
"""Background state shared by the web dashboard and the CLI: account discovery, API polling,
real-time rollout tailing, and the user-triggered actions."""
from __future__ import annotations

import collections
import threading
import time
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Dict, List, Optional

from . import __version__, browsers, paths, store, usage
from .identity import fmt_local, identity, now_utc, to_json_value
from .oauth import BrowserLogin, DeviceCodeLogin, OAuthError, apply_refreshed, refresh_tokens
from .store import StoreError
from .usage import AccountState, RolloutTailer, UsageError


class Monitor:
    def __init__(self, interval: int = 30, credits_max_age: int = 180):
        self.interval = max(10, int(interval))
        self.credits_max_age = credits_max_age
        self.lock = threading.RLock()
        self.states: Dict[str, AccountState] = {}
        self.order: List[str] = []
        self.last_refresh: Optional[datetime] = None
        self.refreshing = False
        self.backoff_until: Optional[datetime] = None
        self.log: collections.deque = collections.deque(maxlen=30)
        self._tailers: Dict[str, RolloutTailer] = {}
        self._main_auth_mtime: Optional[datetime] = None
        self._signature = ""
        self._credits_fetched: Dict[str, datetime] = {}
        self._auto_refresh_at: Dict[str, datetime] = {}
        self._stop = threading.Event()
        self._threads: List[threading.Thread] = []
        self.login: Optional[object] = None
        self.login_name: Optional[str] = None
        self.login_mode: Optional[str] = None
        self.login_relogin = False

    # ------------------------------------------------------------ logging

    def note(self, message: str) -> None:
        stamp = datetime.now().strftime("%H:%M:%S")
        with self.lock:
            self.log.appendleft(f"{stamp} {message}")

    # ------------------------------------------------------------ profiles

    def _signature_now(self) -> str:
        parts = []
        for p in ([store.main_profile()] if store.main_profile() else []) + store.profiles():
            try:
                st = p.auth_path.stat()
                parts.append(f"{p.auth_path}|{st.st_mtime}|{st.st_size}")
            except OSError:
                parts.append(f"{p.auth_path}|missing")
        return "\n".join(parts)

    def reload_profiles(self) -> None:
        msg = store.adopt()
        if msg:
            self.note(msg)
        main = store.main_profile()
        profs = store.profiles()
        main_account = main.account_id if main else ""
        with self.lock:
            previous = {s.ident.get("account_id"): s for s in self.states.values() if s.ident.get("account_id")}
            new_states: Dict[str, AccountState] = {}
            order: List[str] = []

            def carry(st: AccountState) -> None:
                acct = st.ident.get("account_id")
                prev = previous.get(acct) if acct else None
                if prev:
                    st.usage, st.credits, st.fetched_at, st.error = prev.usage, prev.credits, prev.fetched_at, prev.error
                    st.live, st.live_extras = prev.live, prev.live_extras
                elif acct:
                    cached = usage.load_cache(acct)
                    if cached:
                        st.usage, st.credits, st.fetched_at = cached
                        st.error = "showing cached data"

            for p in profs:
                st = AccountState(name=p.name, display_name=p.display_name, active=bool(main_account and p.account_id == main_account),
                                  is_main=False, auth_path=str(p.auth_path), ident=p.ident)
                carry(st)
                new_states[p.name] = st
                order.append(p.name)
            if main and main.auth and not any(p.account_id == main_account for p in profs):
                st = AccountState(name="main", display_name=main.display_name, active=True, is_main=True,
                                  auth_path=str(main.auth_path), ident=main.ident)
                carry(st)
                new_states["main"] = st
                order.insert(0, "main")
            order.sort(key=lambda n: (not new_states[n].active, n.lower()))
            self.states, self.order = new_states, order
            self._main_auth_mtime = main.file_mtime() if main else None
            self._signature = self._signature_now()

    def _auth_for(self, name: str) -> dict:
        st = self.states[name]
        data = store.read_json(Path(st.auth_path))
        if not data:
            raise StoreError(f"cannot read {st.auth_path}")
        return data

    # ------------------------------------------------------------ fetching

    def refresh_all(self, force: bool = False) -> None:
        from . import demo
        if demo.enabled():
            return
        with self.lock:
            if self.refreshing:
                return
            if not force and self.backoff_until and self.backoff_until > now_utc():
                return
            self.refreshing = True
            names = list(self.order)
        try:
            threads = [threading.Thread(target=self.refresh_one, args=(n,), daemon=True) for n in names]
            for t in threads:
                t.start()
            for t in threads:
                t.join(timeout=40)
        finally:
            with self.lock:
                self.refreshing = False
                self.last_refresh = now_utc()

    def refresh_one(self, name: str) -> None:
        with self.lock:
            st = self.states.get(name)
            if not st:
                return
            auth = store.read_json(Path(st.auth_path))
            acct = st.ident.get("account_id") or name
            need_credits = st.credits is None or (now_utc() - self._credits_fetched.get(acct, datetime.min.replace(tzinfo=timezone.utc))).total_seconds() > self.credits_max_age
        if not auth:
            with self.lock:
                st.error = "auth.json unreadable"
            return
        # at most one automatic refresh-token exchange per account per 10 minutes
        last_try = self._auto_refresh_at.get(acct)
        allow_refresh = last_try is None or (now_utc() - last_try).total_seconds() > 600
        try:
            u, auth, refreshed = usage.fetch_usage_auto(Path(st.auth_path), also_main=st.active and not st.is_main,
                                                        allow_refresh=allow_refresh)
            if refreshed:
                self._auto_refresh_at[acct] = now_utc()
                self.note(f"token for {st.display_name} was rejected; refreshed it automatically (like Codex does)")
                st.ident = identity(auth)
            credits = usage.fetch_reset_credits(auth) if need_credits else None
            fetched = now_utc()
            with self.lock:
                st.usage, st.fetched_at, st.error = u, fetched, None
                if credits is not None:
                    st.credits = credits
                    self._credits_fetched[acct] = fetched
            usage.save_cache(st.ident.get("account_id", ""), u, st.credits, fetched)
        except UsageError as e:
            with self.lock:
                st.error = str(e)
                if e.status == 401 and allow_refresh:
                    self._auto_refresh_at[acct] = now_utc()
                if e.status == 429:
                    self.backoff_until = now_utc() + timedelta(seconds=120)
                    self.note("HTTP 429: pausing automatic refresh for 2 minutes")

    # ------------------------------------------------------------ live events

    def poll_live(self) -> None:
        now = now_utc()
        with self.lock:
            items = [(n, s) for n, s in self.states.items()]
            main_mtime = self._main_auth_mtime
        for name, st in items:
            dirs = []
            if st.active:
                dirs.append((paths.codex_home() / "sessions", True))
            if not st.is_main:
                own = Path(st.auth_path).parent / "sessions"
                if own.is_dir():
                    dirs.append((own, False))
            for d, is_main_dir in dirs:
                tailer = self._tailers.get(str(d))
                if tailer is None:
                    tailer = RolloutTailer(d)
                    self._tailers[str(d)] = tailer
                main_ev, extras = tailer.poll_latest(now)
                with self.lock:
                    if main_ev and not (is_main_dir and main_mtime and main_ev.timestamp < main_mtime):
                        if st.live is None or main_ev.timestamp > st.live.timestamp:
                            st.live = main_ev
                            used = (main_ev.primary or main_ev.secondary or {}).get("used_percent")
                            self.note(f"live event for {st.display_name}: used {used}% at {fmt_local(main_ev.timestamp, '%H:%M:%S')}")
                    for lid, ev in extras.items():
                        if is_main_dir and main_mtime and ev.timestamp < main_mtime:
                            continue
                        prev = st.live_extras.get(lid)
                        if prev is None or ev.timestamp > prev.timestamp:
                            st.live_extras[lid] = ev

    # ------------------------------------------------------------ background

    def start_background(self) -> None:
        from . import demo
        if demo.enabled():
            with self.lock:
                self.states = {s.name: s for s in demo.states()}
                self.order = list(self.states.keys())
                self.last_refresh = now_utc()
            self.note("demo mode: fabricated accounts, no network")
            return
        self.reload_profiles()
        threading.Thread(target=self.refresh_all, daemon=True).start()

        def refresher() -> None:
            while not self._stop.wait(self.interval):
                self.refresh_all()

        def live() -> None:
            while not self._stop.wait(2):
                try:
                    self.poll_live()
                except Exception as e:  # never let the tailer kill the loop
                    self.note(f"live tail error: {e}")

        def files() -> None:
            while not self._stop.wait(10):
                try:
                    if self._signature_now() != self._signature:
                        self.note("auth.json changed on disk; reloading accounts")
                        self.reload_profiles()
                        threading.Thread(target=self.refresh_all, daemon=True).start()
                except Exception as e:
                    self.note(f"file poll error: {e}")

        for fn in (refresher, live, files):
            t = threading.Thread(target=fn, daemon=True)
            t.start()
            self._threads.append(t)
        try:
            self.poll_live()
        except Exception:
            pass

    def stop(self) -> None:
        self._stop.set()

    # ------------------------------------------------------------ actions

    def switch(self, name: str) -> List[str]:
        msgs = store.activate(name)
        for m in msgs:
            self.note(m)
        self.reload_profiles()
        threading.Thread(target=self.refresh_all, kwargs={"force": True}, daemon=True).start()
        return msgs

    def save_main(self, name: str) -> str:
        p = store.save_main(name)
        self.note(f"saved current login as account '{p.name}'")
        self.reload_profiles()
        return p.name

    def auth_text(self, name: str) -> str:
        st = self.states[name]
        return Path(st.auth_path).read_text(encoding="utf-8")

    def refresh_token(self, name: str) -> str:
        auth = self._auth_for(name)
        rt = (auth.get("tokens") or {}).get("refresh_token")
        if not rt:
            raise StoreError("this account has no refresh_token")
        new = refresh_tokens(rt)
        st = self.states[name]
        apply_refreshed(Path(st.auth_path), new)
        if st.active and not st.is_main:
            apply_refreshed(paths.main_auth_path(), new)
        self.note(f"refreshed tokens for {st.display_name}")
        self.reload_profiles()
        threading.Thread(target=self.refresh_all, kwargs={"force": True}, daemon=True).start()
        return "refreshed"

    def start_login(self, name: str, mode: str = "browser", relogin: bool = False) -> dict:
        if self.login is not None and getattr(self.login, "phase", "") in ("starting", "waiting", "exchanging"):
            raise StoreError("a login is already in progress")
        if relogin:
            st = self.states.get(name)
            if not st:
                raise StoreError(f"unknown account {name!r}")
            directory = Path(st.auth_path).parent
        else:
            name = store.sanitize(name)
            directory = paths.accounts_dir() / name
            if (directory / "auth.json").exists():
                raise StoreError(f"account '{name}' already exists")
        flow = BrowserLogin(directory) if mode == "browser" else DeviceCodeLogin(directory)
        self.login, self.login_name, self.login_mode, self.login_relogin = flow, name, mode, relogin
        flow.start()
        self.note(f"login started for '{name}' ({mode})")
        return self.login_status()

    def login_status(self) -> dict:
        flow = self.login
        if flow is None:
            return {"phase": "idle"}
        out = {"phase": flow.phase, "name": self.login_name, "mode": self.login_mode, "error": flow.error,
               "url": getattr(flow, "url", None), "code": getattr(flow, "code", None)}
        ident = getattr(flow, "result_identity", None)
        if ident:
            out["email"] = ident.get("email")
            out["plan"] = ident.get("plan")
        return out

    def open_login_url(self, private: bool) -> str:
        flow = self.login
        url = getattr(flow, "url", None) if flow else None
        if not url:
            raise StoreError("no login in progress")
        if private:
            name = browsers.open_private(url)
            if name:
                return name
        return "default browser" if browsers.open_default(url) else "none"

    def cancel_login(self) -> None:
        if self.login is not None:
            self.login.cancel()
            self.note("login cancelled")

    def finish_login_if_done(self) -> None:
        """Called by the web layer after a successful login so the new account shows up."""
        flow = self.login
        if flow is not None and flow.phase == "success":
            self.reload_profiles()
            threading.Thread(target=self.refresh_all, kwargs={"force": True}, daemon=True).start()

    # ------------------------------------------------------------ snapshot

    def snapshot(self) -> dict:
        now = now_utc()
        with self.lock:
            accounts = [self.states[n].view(now) for n in self.order]
            data = {
                "version": __version__,
                "now": now,
                "interval": self.interval,
                "last_refresh": self.last_refresh,
                "refreshing": self.refreshing,
                "accounts_dir": paths.display_path(paths.accounts_dir()),
                "codex_home": str(paths.codex_home()),
                "private_browser": (browsers.find_private_browser() or (None,))[0],
                "accounts": accounts,
                "login": self.login_status(),
                "log": list(self.log),
            }
        return to_json_value(data)
